Add OpenPMD support (#1050)
parthenon-hpc-lab/parthenon: doc/sphinx/src/outputs.rst

Outputs from Parthenon are controlled via ``<parthenon/output*>`` blocks,
where ``*`` should be replaced by a unique integer for each block.

Various output types are supported for writing both raw data (e.g., in

> openPMD/ADIOS2/BP5 or HDF5 format) andprocessed data (like high level aggregated

Reviewer: ```suggestion
openPMD/ADIOS2/BP5 or HDF5 format) and processed data (like high level aggregated
```

> history files or histograms), see following subsections for more details.

Reviewer: ```suggestion
history files or histograms), see the following subsections for more details.
```

While HDF5 has been the default in the past, openPMD/ADIOS2/BP5 is expected to

> become the default in the near term future given its features, flexiblity,

Reviewer: ```suggestion
become the default in the near-term future given its features, flexiblity,
```

unaffected and the ``final`` and ``now`` files will be overwritten each
time without warning.

openPMD
-------

Parthenon supports writing outputs in the "Open Standard for Particle-Mesh Data",
`openPMD <https://www.openpmd.org>`__.
While openPMD supports various backends only the
`ADIOS2 <https://adios2.readthedocs.io/en/v2.11.0/>`__ IO framework is currently
supported/tested with files being writen in BP5 format.

The most simple output block for openPMD output is

::

   <parthenon/output6>
   file_type   = openpmd    # write data in openPMD format
   output_type = restart
   dt          = 0.125      # time increment between outputs

and will produce outputs in directories named ``parthenon.out6.#####.bp``
where the ``6`` (in the output block header is arbitrary and used
as default value in the resulting directory name) and ``#####`` an increasing
zero-padded integer that is increased for each output written.

> Restarting from outputs using these minimal, simple block is supported by

Reviewer: ```suggestion
Restarting from outputs using this minimal, simple block is supported by
```

default (i.e., it contains all the information for restarting a simulation,
such as independent fields including particles, mesh structure, input file,
and other parameters).

More complex configurations are possible, e.g.,

::

   <parthenon/output5>
   file_type   = openpmd           # write data in openPMD format
   dt          = 0.125             # time increment between outputs
   id          = compressed        # arb. string used in resulting directory name
   backend_config = openpmd.toml   # path to external config file controlling IO behavior

>    output_type = data              # only write subset of fields

Reviewer 2: this is a good idea. Wonder if we should make hdf5 work like this... maybe too late.

Author: I think the change would make logical sense, but I'm not sure there's a clean, fully backwards compatible approach.

Reviewer 2: Yeah...

   variables   = density, pressure # list of fields to be written in output

   swarms = tracers, photons       # Particle swarms
   swarm_variables = x, y, z       # swarm variables output for every swarm

>    # Each swarm can sepcify in a separate list which additional

Reviewer: ```suggestion
   # Each swarm can specify in a separate list which additional
```

   # variables it would like to output.
   tracers_variables = x, y, z, rho, id
   photons_variables = x, y, z, frequency

Here, the resulting outputs will be written to ``parthenon.compressed.#####.bp``
directories and only include the mesh fields named ``density`` and ``pressure``
and particle species ``tracers`` and ``photons`` with their respective fields.
Moreover, an external config file (arbitrarily named ``openpmd.toml``)

> is parsed to tune IO behavior (like performance or compression).

Reviewer: Is config file in a Parthenon specific format or is it something that OpenPMD supports directly?

Is there any way to do this tuning directly in an input file? Not sure that we want that, but it was a thought that came to mind.

Reviewer: Oh, I think this question is addressed below.

Author: Yes, this is all external and can be configured at runtime via an external file.

================================

Native Parthenon-based
----------------------

A rudimentary postprocessing option is available to analyze data in
restart files within the downstream/Parthenon framework, i.e., making use of
native compute capabilities.

To trigger this kind of analysis, restart with ``-a /PATH/TO/RESTART/OUTPUT``
(instead of ``-r``).
This will launch the standard driver as if a simulation would be restarted,
but never call the main time loop.
Only the callbacks ``UserWorkBeforeLoop`` are executed.
Afterwards, all output blocks that have the parameter ``analysis_output=true`` are
going to be processed including ``UserWorkBeforeOutput`` callbacks.

Note, the standard modifications to the original input parameters via the command line
or via an input file apply.
A typical use case is, for example, to calculate histograms a posteriori, i.e., a new
output block is specified in a file called ``sample_hist.in`` with all necessary details
(particularly the ``analysis_output=true`` parameter within said block) and then the
data can be processed via ``-a /PATH/TO/RESTART/OUTPUT -i sample_hist.in``.

Python scripts
--------------

openPMD outputs
^^^^^^^^^^^^^^^

> Outputs written following the openPMD standard can be processed by any tool implemting

Reviewer: ```suggestion
Outputs written following the openPMD standard can be processed by any tool implementing
```
